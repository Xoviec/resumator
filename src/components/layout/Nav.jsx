import React, { useContext } from "react";
import { Link, useHistory } from "react-router-dom";
import { makeStyles } from "@material-ui/core/styles";
import {
  AppBar,
  IconButton,
  Menu,
  MenuItem,
  Toolbar,
  Avatar,
  Divider,
} from "@material-ui/core";
import { AccountCircle } from "@material-ui/icons";
import "firebase/firestore";
import "firebase/auth";
import frontmenLogo from "../../assets/svg/frontmen-logo.svg";
import { FirebaseAppContext } from "../../context/FirebaseContext";

const useStyles = makeStyles((theme) => ({
  appBar: {
    zIndex: theme.zIndex.drawer + 1,
  },
  grow: {
    flexGrow: 1,
  },
  logo: {
    width: 40,
    height: 40,
  },
  linkItem: {
    color: theme.palette.primary.main,
    textDecoration: "none",
  },
}));

const Nav = () => {
  const { firebase, user } = useContext(FirebaseAppContext);

  const history = useHistory();
  const goTo = (path) => history.push(path);

  const classes = useStyles();
  const [anchorEl, setAnchorEl] = React.useState(null);

  const isMenuOpen = Boolean(anchorEl);

  const handleProfileMenuOpen = (event) => {
    setAnchorEl(event.currentTarget);
  };

  const handleMenuClose = () => {
    setAnchorEl(null);
  };

  const signOut = async () => {
    await firebase.auth().signOut();
    goTo("/");
  };

  const adminMenuItems = (
    <div>
      <MenuItem>
        <Link to="/new" className={classes.linkItem}>
          Add Resume
        </Link>
      </MenuItem>
      <MenuItem>
        <Link to="/skills" className={classes.linkItem}>
          Manage Skills
        </Link>
      </MenuItem>
      <Divider />
    </div>
  );

  const menuId = "primary-search-account-menu";
  const renderMenu = (
    <Menu
      anchorEl={anchorEl}
      anchorOrigin={{ vertical: "top", horizontal: "right" }}
      id={menuId}
      keepMounted
      transformOrigin={{ vertical: "top", horizontal: "right" }}
      open={isMenuOpen}
      onClose={handleMenuClose}
    >
      {user?.userRec?.isManager && adminMenuItems}
      <MenuItem onClick={signOut}>Sign out</MenuItem>
    </Menu>
  );

  const avatarComponent =
    user && user.photoURL ? (
      <Avatar alt={user.displayName} src={user.photoURL} />
    ) : (
      <AccountCircle />
    );

  return (
    <div className={classes.navContainer}>
      <div className={classes.grow}>
        <AppBar position="fixed" className={classes.appBar}>
          <Toolbar>
            <a href="/" title="Home">
              <img className={classes.logo} src={frontmenLogo} alt="logo" />
            </a>
            <div className={[classes.grow].join(" ")} />
            <div className={classes.grow} />
            <IconButton
              edge="end"
              aria-label="account of current user"
              aria-controls={menuId}
              aria-haspopup="true"
              onClick={handleProfileMenuOpen}
              color="inherit"
            >
              {avatarComponent}
            </IconButton>
          </Toolbar>
        </AppBar>
      </div>
      {renderMenu}
    </div>
  );
};

export default Nav;
