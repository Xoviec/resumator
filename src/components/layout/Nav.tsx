import React from "react";
import { Link, useHistory } from "react-router-dom";
import { makeStyles } from "@material-ui/core/styles";
import {
  AppBar,
  IconButton,
  Menu,
  MenuItem,
  Toolbar,
  Avatar,
  Divider,
} from "@material-ui/core";
import { AccountCircle } from "@material-ui/icons";
import "firebase/firestore";
import "firebase/auth";
import frontmenLogo from "../../assets/svg/frontmen-logo.svg";
import { useFirebaseApp } from "../../context/FirebaseContext";

const useStyles = makeStyles((theme) => ({
  appBar: {
    zIndex: theme.zIndex.drawer + 1,
  },
  grow: {
    flexGrow: 1,
  },
  logo: {
    width: 40,
    height: 40,
  },
  linkItem: {
    color: theme.palette.primary.main,
    textDecoration: "none",
  },
}));

const Nav: React.FC = () => {
  const { firebase, authUser, userRecord } = useFirebaseApp();

  const history = useHistory();
  const goTo = (path: string) => history.push(path);

  const classes = useStyles();
  const [anchorEl, setAnchorEl] = React.useState<HTMLElement | null>(null);

  const isMenuOpen = Boolean(anchorEl);

  const handleProfileMenuOpen = (
    event: React.MouseEvent<HTMLButtonElement, MouseEvent>
  ) => {
    setAnchorEl(event.currentTarget);
  };

  const handleMenuClose = () => {
    setAnchorEl(null);
  };

  const signOut = async () => {
    await firebase.auth().signOut();
    goTo("/");
  };

  const adminMenuItems = userRecord?.isManager ? (
    <div>
      <MenuItem>
        <Link to="/new" className={classes.linkItem}>
          Add Resume
        </Link>
      </MenuItem>
      <MenuItem>
        <Link to="/skills" className={classes.linkItem}>
          Manage Skills
        </Link>
      </MenuItem>
      <Divider />
    </div>
  ) : null;

  const menuId = "primary-search-account-menu";
  const renderMenu = (
    <Menu
      anchorEl={anchorEl}
      anchorOrigin={{ vertical: "top", horizontal: "right" }}
      id={menuId}
      keepMounted
      transformOrigin={{ vertical: "top", horizontal: "right" }}
      open={isMenuOpen}
      onClose={handleMenuClose}
    >
      {adminMenuItems}
      <MenuItem onClick={signOut}>Sign out</MenuItem>
    </Menu>
  );

  const avatarComponent = authUser?.photoURL ? (
    <Avatar alt={authUser.displayName || "User Avatar"} src={authUser.photoURL} />
  ) : (
    <AccountCircle />
  );

  return (
    <div>
      <div className={classes.grow}>
        <AppBar position="fixed" className={classes.appBar}>
          <Toolbar>
            <a href="/" title="Home">
              <img className={classes.logo} src={frontmenLogo} alt="logo" />
            </a>
            <div className={[classes.grow].join(" ")} />
            <div className={classes.grow} />
            <IconButton
              edge="end"
              aria-label="account of current user"
              aria-controls={menuId}
              aria-haspopup="true"
              onClick={handleProfileMenuOpen}
              color="inherit"
            >
              {avatarComponent}
            </IconButton>
          </Toolbar>
        </AppBar>
      </div>
      {renderMenu}
    </div>
  );
};

export default Nav;
