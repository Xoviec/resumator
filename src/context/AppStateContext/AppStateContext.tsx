import {
  createContext,
  Dispatch,
  FunctionComponent,
  SetStateAction,
  useContext,
  useMemo,
  useState,
} from "react";

export type AppStateContextType = {
  isDrawerOpen: boolean;
  setIsDrawerOpen: Dispatch<SetStateAction<boolean>>;
};

const AppStateContext = createContext<AppStateContextType | undefined>(undefined);

export const useAppState = (): AppStateContextType => {
  const context = useContext(AppStateContext);
  if (context === undefined) {
    throw new Error("useAppState must be used within a AppStateProvider");
  }
  return context;
};

export const AppStateContextProvider: FunctionComponent = ({ children }) => {
  const [isDrawerOpen, setIsDrawerOpen] = useState(false);

  const contextState = useMemo(
    () => ({
      isDrawerOpen,
      setIsDrawerOpen,
    }),
    [isDrawerOpen, setIsDrawerOpen]
  );

  return (
    <AppStateContext.Provider value={contextState}>
      {children}
    </AppStateContext.Provider>
  );
};
