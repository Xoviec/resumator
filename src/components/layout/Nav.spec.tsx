import { act, render } from "@testing-library/react";
import { BrowserRouter } from "react-router-dom";
import { createMemoryHistory } from "history";
import userEvent from "@testing-library/user-event";
import { createTheme } from "@mui/material/styles";
import { ThemeProvider } from "@mui/material";
import { FunctionComponent } from "react";
import {
  useFirebaseApp,
  FirebaseAppContextType,
} from "../../context/FirebaseContext/FirebaseContext";
import { useAppState } from "../../context/AppStateContext/AppStateContext";
import { mocked } from "jest-mock";
import { Nav } from "./Nav";

const ThemeProviderWrapper: FunctionComponent = ({ children }) => {
  const theme = createTheme({});
  return <ThemeProvider theme={theme}>{children}</ThemeProvider>;
};

jest.mock("../../context/FirebaseContext/FirebaseContext");
jest.mock("../../context/AppStateContext/AppStateContext");

beforeEach(jest.clearAllMocks);

describe("Nav", () => {
  it("opens the drawer when clicked on open drawer button", async () => {
    let isDrawerOpen = false;
    const setIsDrawerOpenMock = jest.fn((fn) => {
      isDrawerOpen = fn(isDrawerOpen);
    });

    mocked(useAppState).mockImplementation(() => ({
      isDrawerOpen,
      setIsDrawerOpen: setIsDrawerOpenMock,
    }));

    mocked(useFirebaseApp).mockImplementation(
      () =>
        ({
          firebase: {},
          userRecord: { isManager: true },
        } as FirebaseAppContextType)
    );

    const { asFragment, getByLabelText } = render(
      <BrowserRouter>
        <Nav />
      </BrowserRouter>
    );

    expect(asFragment()).toMatchSnapshot();

    await act(async () => {
      userEvent.click(getByLabelText("open drawer"));
    });

    expect(isDrawerOpen).toEqual(true);
  });

  it("closes the drawer when clicked on close drawer button", async () => {
    let isDrawerOpen = true;
    const setIsDrawerOpenMock = jest.fn((fn) => {
      isDrawerOpen = fn(isDrawerOpen);
    });

    mocked(useAppState).mockImplementation(() => ({
      isDrawerOpen,
      setIsDrawerOpen: setIsDrawerOpenMock,
    }));

    mocked(useFirebaseApp).mockImplementation(
      () =>
        ({
          firebase: {},
          userRecord: { isManager: true },
        } as FirebaseAppContextType)
    );

    const { asFragment, getByLabelText } = render(
      <BrowserRouter>
        <Nav />
      </BrowserRouter>
    );

    expect(asFragment()).toMatchSnapshot();

    await act(async () => {
      userEvent.click(getByLabelText("close drawer"));
    });

    expect(isDrawerOpen).toEqual(false);
  });

  it("signs out and navigates to home when clicked to sign out button", async () => {
    const signOutMock = jest.fn();

    mocked(useAppState).mockImplementation(() => ({
      isDrawerOpen: true,
      setIsDrawerOpen: jest.fn(),
    }));

    mocked(useFirebaseApp).mockImplementation(
      () =>
        ({
          firebase: {
            auth: () => ({
              signOut: signOutMock,
            }),
          },
          // eslint-disable-next-line @typescript-eslint/no-explicit-any
        } as any)
    );

    const memoryHistory = createMemoryHistory();

    const { getByLabelText } = render(
      <BrowserRouter>
        <Nav />
      </BrowserRouter>
    );

    await act(async () => {
      userEvent.click(getByLabelText(/sign out/i));
    });

    expect(signOutMock).toHaveBeenCalled();
    expect(memoryHistory.location.pathname).toEqual("/");
  });

  it("shows adminMenuItems when the user is manager", async () => {
    mocked(useAppState).mockImplementation(() => ({
      isDrawerOpen: true,
      setIsDrawerOpen: jest.fn(),
    }));

    mocked(useFirebaseApp).mockImplementation(
      () =>
        ({
          firebase: {},
          userRecord: { isManager: true },
        } as FirebaseAppContextType)
    );

    const { getByText } = render(
      <ThemeProviderWrapper>
        <BrowserRouter>
          <Nav />
        </BrowserRouter>
      </ThemeProviderWrapper>
    );

    expect(getByText(/add resume/i)).toBeInTheDocument();
    expect(getByText(/manage skills/i)).toBeInTheDocument();
  });

  it("hides adminMenuItems when the user is not manager", () => {
    mocked(useAppState).mockImplementation(() => ({
      isDrawerOpen: true,
      setIsDrawerOpen: jest.fn(),
    }));

    mocked(useFirebaseApp).mockImplementation(
      () =>
        ({
          firebase: {},
          userRecord: { isManager: false },
        } as FirebaseAppContextType)
    );

    const { queryByText } = render(
      <ThemeProviderWrapper>
        <BrowserRouter>
          <Nav />
        </BrowserRouter>
      </ThemeProviderWrapper>
    );

    expect(queryByText(/add resume/i)).not.toBeInTheDocument();
    expect(queryByText(/manage skills/i)).not.toBeInTheDocument();
  });

  it("shows default avatar when user photo not present", () => {
    mocked(useAppState).mockImplementation(() => ({
      isDrawerOpen: true,
      setIsDrawerOpen: jest.fn(),
    }));

    mocked(useFirebaseApp).mockImplementation(
      () =>
        ({
          firebase: {},
        } as FirebaseAppContextType)
    );

    const { getByLabelText } = render(
      <ThemeProviderWrapper>
        <BrowserRouter>
          <Nav />
        </BrowserRouter>
      </ThemeProviderWrapper>
    );

    expect(getByLabelText("user avatar")).toBeInTheDocument();
  });
});
