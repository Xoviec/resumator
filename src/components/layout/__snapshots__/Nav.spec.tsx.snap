// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Nav closes the drawer when clicked on close drawer button 1`] = `
<DocumentFragment>
  <div
    class="MuiBox-root css-0"
  >
    <div
      class="MuiBox-root css-0"
    >
      <header
        class="MuiPaper-root MuiPaper-elevation MuiPaper-elevation4 MuiAppBar-root MuiAppBar-colorPrimary MuiAppBar-positionFixed mui-fixed css-ivnu1y-MuiPaper-root-MuiAppBar-root"
      >
        <div
          class="MuiToolbar-root MuiToolbar-gutters MuiToolbar-regular css-1t29gy6-MuiToolbar-root"
        >
          <button
            aria-label="close drawer"
            class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit MuiIconButton-edgeStart MuiIconButton-sizeMedium css-1mbzpyo-MuiButtonBase-root-MuiIconButton-root"
            tabindex="0"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="ChevronLeftIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M15.41 7.41 14 6l-6 6 6 6 1.41-1.41L10.83 12z"
              />
            </svg>
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
          <a
            class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-78trlr-MuiButtonBase-root-MuiIconButton-root"
            href="/"
            tabindex="0"
          >
            <img
              alt="iO Logo white"
              height="44"
              src="iO-logo-white.png"
            />
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </a>
          <div
            class="MuiTypography-root MuiTypography-h6 css-5949ms-MuiTypography-root"
          >
            Resumator
          </div>
          <button
            aria-controls="primary-search-account-menu"
            aria-haspopup="true"
            aria-label="account of current user"
            class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit MuiIconButton-edgeEnd MuiIconButton-sizeMedium css-1ei1svh-MuiButtonBase-root-MuiIconButton-root"
            tabindex="0"
            type="button"
          >
            <svg
              aria-hidden="true"
              aria-label="user avatar"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="AccountCircleIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 3c1.66 0 3 1.34 3 3s-1.34 3-3 3-3-1.34-3-3 1.34-3 3-3zm0 14.2c-2.5 0-4.71-1.28-6-3.22.03-1.99 4-3.08 6-3.08 1.99 0 5.97 1.09 6 3.08-1.29 1.94-3.5 3.22-6 3.22z"
              />
            </svg>
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
        </div>
      </header>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Nav opens the drawer when clicked on open drawer button 1`] = `
<DocumentFragment>
  <div
    class="MuiBox-root css-0"
  >
    <div
      class="MuiBox-root css-0"
    >
      <header
        class="MuiPaper-root MuiPaper-elevation MuiPaper-elevation4 MuiAppBar-root MuiAppBar-colorPrimary MuiAppBar-positionFixed mui-fixed css-ivnu1y-MuiPaper-root-MuiAppBar-root"
      >
        <div
          class="MuiToolbar-root MuiToolbar-gutters MuiToolbar-regular css-1t29gy6-MuiToolbar-root"
        >
          <button
            aria-label="open drawer"
            class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit MuiIconButton-edgeStart MuiIconButton-sizeMedium css-1mbzpyo-MuiButtonBase-root-MuiIconButton-root"
            tabindex="0"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="MenuIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M3 18h18v-2H3v2zm0-5h18v-2H3v2zm0-7v2h18V6H3z"
              />
            </svg>
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
          <a
            class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-78trlr-MuiButtonBase-root-MuiIconButton-root"
            href="/"
            tabindex="0"
          >
            <img
              alt="iO Logo white"
              height="44"
              src="iO-logo-white.png"
            />
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </a>
          <div
            class="MuiTypography-root MuiTypography-h6 css-5949ms-MuiTypography-root"
          >
            Resumator
          </div>
          <button
            aria-controls="primary-search-account-menu"
            aria-haspopup="true"
            aria-label="account of current user"
            class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit MuiIconButton-edgeEnd MuiIconButton-sizeMedium css-1ei1svh-MuiButtonBase-root-MuiIconButton-root"
            tabindex="0"
            type="button"
          >
            <svg
              aria-hidden="true"
              aria-label="user avatar"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="AccountCircleIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 3c1.66 0 3 1.34 3 3s-1.34 3-3 3-3-1.34-3-3 1.34-3 3-3zm0 14.2c-2.5 0-4.71-1.28-6-3.22.03-1.99 4-3.08 6-3.08 1.99 0 5.97 1.09 6 3.08-1.29 1.94-3.5 3.22-6 3.22z"
              />
            </svg>
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
        </div>
      </header>
    </div>
  </div>
</DocumentFragment>
`;
